src/algebra/module/torsion.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import algebra.module
+import linear_algebra.quotient
+import ring_theory.ideal.quotient
+
+/-!
+# Torsion submodules
+
+## Main definitions
+
+* `torsion R M a` : the `a`-torsion submodule, containing all elements `x` of `M` such that
+  `a • x = 0`.
+* `torsion' R M` : the torsion submoule, containing all elements `x` of `M` such that  `a • x = 0`
+  for some non-zero `a` in `R`.
+
+## Main statements
+
+* `torsion' R M` is a submodule when `R` is a domain.
+* `torsion R M a` can be viewed as a `(R ⧸ R·a)`-module.
+* `quot_torsion_is_torsion_free` : `(M ⧸ torsion' R M)` is a torsion-free module, that is, there is
+  no non-zero `a`, `x` such that `a • x = 0`.
+
+## Notation
+
+* The notions are defined for a `comm_semiring R` and a `module R M`. Some additional hypotheses on
+  `R` and `M`are required by some lemmas.
+* The letters `a`, `b`, ... are used for scalars (in `R`), while `x`, `y`, ... are used for vectors
+  (in `M`).
+
+## Tags
+
+Torsion, submodule, module, quotient
+-/
+
+
+section defs
+variables (R M : Type*) [comm_semiring R] [add_comm_monoid M] [module R M] (a : R)
+
+/-- The `a`-torsion submodule, for `a` in `R` -/
+def torsion : submodule R M := (distrib_mul_action.to_linear_map _ _ a).ker
+
+/-- The torsion submodule, only defined when `R` is a domain. -/
+def torsion' [no_zero_divisors R] [nontrivial R] : submodule R M :=
+{ carrier := { x | ∃ a : R, a • x = 0 ∧ a ≠ 0 },
+  zero_mem' := ⟨1, smul_zero _, one_ne_zero⟩,
+  add_mem' := λ x y ⟨a, hx, ha⟩ ⟨b, hy, hb⟩,
+    ⟨b * a,
+      by rw [smul_add, ← smul_smul, mul_comm, ← smul_smul, hx, hy, smul_zero, smul_zero, add_zero],
+      mul_ne_zero hb ha⟩,
+  smul_mem' := λ a x ⟨b, h, hb⟩, ⟨b, by rw [smul_smul, mul_comm, ← smul_smul, h, smul_zero], hb⟩ }
+end defs
+
+section
+variables {R M : Type*} [comm_semiring R] [add_comm_monoid M] [module R M] (a : R)
+
+@[simp] lemma smul_torsion (x : torsion R M a) : a • x = 0 := subtype.ext x.prop
+
+/-- A module is torsion-free (`no_zero_smul_divisors`) iff its torsion submodule is trivial. -/
+lemma no_zero_smul_divisors_iff_torsion_bot [no_zero_divisors R] [nontrivial R] :
```

Comment:
```suggestion
lemma no_zero_smul_divisors_iff_torsion'_bot [no_zero_divisors R] [nontrivial R] :
```

I think it might be helpful to do all the `torsion` lemmas in a different half of the file to the `torsion'` ones.